planner/core/push_down_predicate.go

```diff
@@ -0,0 +1,249 @@
+// Copyright 2023 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package core
+
+import (
+	"sort"
+
+	"github.com/pingcap/tidb/expression"
+	"github.com/pingcap/tidb/kv"
+	"github.com/pingcap/tidb/parser/ast"
+	"github.com/pingcap/tidb/sessionctx"
+	"github.com/pingcap/tidb/util/logutil"
+	"go.uber.org/zap"
+)
+
+const (
+	// selectivity = (row count after filter) / (row count before filter), smaller is better
+	selectivityThreshold = 0.7
+	// income = (1 - selectivity) * restColumnCount, greater is better
+	incomeImproveThreshold = 0.1
+)
+
+// expressionGroup is used to store
+// 1. a group of expressions
+// 2. the selectivity of the expressions
+type expressionGroup struct {
+	exprs       []expression.Expression
+	selectivity float64
+}
+
+func predicatePushDownToTableScan(sctx sessionctx.Context, plan PhysicalPlan) {
+	switch p := plan.(type) {
+	case *PhysicalSelection:
+		physicalTableScan, ok := plan.Children()[0].(*PhysicalTableScan)
+		if ok && physicalTableScan.isFullScan() && physicalTableScan.StoreType == kv.TiFlash {
+			// Only when the table scan is a full scan and the store type is TiFlash,
+			// we will try to push down predicates.
+			predicatePushDownToTableScanImpl(sctx, p, physicalTableScan)
+		}
+	case *PhysicalTableReader:
+		predicatePushDownToTableScan(sctx, p.tablePlan)
+	default:
+		for _, child := range plan.Children() {
+			predicatePushDownToTableScan(sctx, child)
+		}
+	}
+}
+
+/**
+ * @brief: This function is used to remove the empty selection in plan
+ * 		   when late materialization is enabled and all conditions are pushed down.
+ * @param: plan: the physical plan
+ */
+func removeEmptySelection(plan PhysicalPlan) {
+	var newChildren []PhysicalPlan
+	removed := false
+	for _, child := range plan.Children() {
+		switch x := child.(type) {
```

Comment:
It seems that all indentation in your code is using 8 spaces. It will look better if you use 4 uniformly as before.

Reply:
Due to incorrect settings in GitHub, fixed.